core/src/smartcontracts/isi/domain.rs

```diff
@@ -391,32 +391,34 @@ pub mod isi {
 /// Query module provides [`Query`] Domain related implementations.
 pub mod query {
     use eyre::Result;
-    use iroha_data_model::{domain::Domain, query::error::QueryExecutionFail as Error};
+    use iroha_data_model::{
+        domain::Domain,
+        query::{
+            error::QueryExecutionFail as Error,
+            predicate::{predicate_atoms::domain::DomainPredicateBox, CompoundPredicate},
+        },
+    };
     use iroha_primitives::json::JsonString;
 
     use super::*;
-    use crate::state::StateReadOnly;
+    use crate::{smartcontracts::ValidIterableQuery, state::StateReadOnly};
 
-    impl ValidQuery for FindAllDomains {
+    impl ValidIterableQuery for FindAllDomains {
         #[metrics(+"find_all_domains")]
         fn execute<'state>(
-            &self,
+            self,
+            filter: CompoundPredicate<DomainPredicateBox>,
             state_ro: &'state impl StateReadOnly,
-        ) -> Result<Box<dyn Iterator<Item = Domain> + 'state>, Error> {
-            Ok(Box::new(state_ro.world().domains_iter().cloned()))
-        }
-    }
-
-    impl ValidQuery for FindDomainById {
-        #[metrics(+"find_domain_by_id")]
-        fn execute(&self, state_ro: &impl StateReadOnly) -> Result<Domain, Error> {
-            let id = &self.id;
-            iroha_logger::trace!(%id);
-            Ok(state_ro.world().domain(id)?.clone())
+        ) -> std::result::Result<impl Iterator<Item = Domain> + 'state, Error> {
+            Ok(state_ro
+                .world()
+                .domains_iter()
+                .filter(move |&v| filter.applies(v))
+                .cloned())
         }
     }
 
-    impl ValidQuery for FindDomainKeyValueByIdAndKey {
+    impl ValidSingularQuery for FindDomainKeyValueByIdAndKey {
```

Comment:
What do you think about replacing this type of queries with `FindDomainMetadata` and then use filters to get specific keys?

Reply:
I am kinda split on this... This won't reduce the number of queries we have, but it will improve flexibility just a little bit..

Though I feel projections are a more general solution for this

Reply:
at the very least let's rename `FindDomainKeyValueByIdAndKey` to `FindDomainMetadata`.
But my question is, if we have `FindAllDomains`, isn't it possible to get rid of this query?

Reply:
> But my question is, if we have FindAllDomains, isn't it possible to get rid of this query?

It is possible, but only if you don't consider the additional traffic generated by this: you need to fetch the full `Domain` object and then query the metadata KV store on the client side. `FindDomainKeyValueByIdAndKey`, on the other hand, does its lookup on the peer's side and only sends the queried metadata value.

Reply:
we can resolve this?